Remove null values returned from available post MIME types

ironprogrammer opened the pull request:
Removes nulls from returned available post MIME types, and adds unit test coverage for `get_available_post_mime_types()` and `pre_get_available_post_mime_types` filter.

Trac ticket: https://core.trac.wordpress.org/ticket/59195

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

ironprogrammer committed:
Remove nulls from post $mime_types

ironprogrammer committed:
Add unit tests for get_available_post_mime_types

ironprogrammer committed:
Add uploads and compare with expected types

ironprogrammer committed:
Clean up uploads when done

ironprogrammer committed:
Remove redundant assertions

anton-vlasenko commented on src/wp-includes/post.php:
```diff
@@ -8073,7 +8073,8 @@ function get_available_post_mime_types( $type = 'attachment' ) {
 		$mime_types = $wpdb->get_col( $wpdb->prepare( "SELECT DISTINCT post_mime_type FROM $wpdb->posts WHERE post_type = %s", $type ) );
```
Empty `post_mime_type` values could probably also be filtered at the SQL query level too:
```suggestion
		$mime_types = $wpdb->get_col( $wpdb->prepare( "SELECT DISTINCT post_mime_type FROM $wpdb->posts WHERE post_type = %s AND post_mime_type <> ''", $type ) );
```

ironprogrammer replied:
This suggestion does more clearly convey that we don't want empty MIMEs returned. The subsequent `array_filter( $mime_types )` overlaps/reinforces this, but encompasses the pre-filter's return value as well.

Updated in f79ce9f47c.

anton-vlasenko commented on tests/phpunit/tests/post/mimeTypes.php:
```diff
@@ -0,0 +1,66 @@
+	 */
+	public function test_get_available_post_mime_types() {
+		// Upload a JPEG image.
+		$filename = DIR_TESTDATA . '/images/test-image.jpg';
+		$contents = file_get_contents( $filename );
+		$upload   = wp_upload_bits( wp_basename( $filename ), null, $contents );
+		$this->assertEmpty( $upload['error'] );
+		$this->_make_attachment( $upload );
+
+		// Upload a PDF file.
+		$filename = DIR_TESTDATA . '/images/test-alpha.pdf';
+		$contents = file_get_contents( $filename );
+		$upload   = wp_upload_bits( wp_basename( $filename ), null, $contents );
+		$this->assertEmpty( $upload['error'] );
+		$this->_make_attachment( $upload );
+
+		$mime_types = get_available_post_mime_types();
+
+		$this->assertSame( array( 'image/jpeg', 'application/pdf' ), $mime_types );
```
Assertions should probably have the `$message` parameter defined as recommended by the guidelines on https://make.wordpress.org/core/handbook/testing/automated-testing/writing-phpunit-tests/#using-assertions

ironprogrammer replied:
Thanks! Addressed in 3a4c4162f0.

anton-vlasenko reviewed (commented):
Please find my code review below. Thank you.

ironprogrammer committed:
Add assertion failure messages

Tests with multiple assertions should include failure status messages.

Props @anton-vlasenko.

ironprogrammer committed:
Filters out empty MIME types from query

Props @anton-vlasenko.

anton-vlasenko commented on src/wp-includes/post.php:
```diff
@@ -8070,10 +8070,11 @@ function get_available_post_mime_types( $type = 'attachment' ) {
 
 	if ( ! is_array( $mime_types ) ) {
-		$mime_types = $wpdb->get_col( $wpdb->prepare( "SELECT DISTINCT post_mime_type FROM $wpdb->posts WHERE post_type = %s", $type ) );
+		$mime_types = $wpdb->get_col( $wpdb->prepare( "SELECT DISTINCT post_mime_type FROM $wpdb->posts WHERE post_type = %s AND post_mime_type != ''", $type ) );
```
It appears that the `<>` operator is included in the ANSI standard, but most SQL databases also support the `!=` operator.
Therefore, there is virtually no difference between the two.

anton-vlasenko commented on src/wp-includes/post.php:
```diff
@@ -8070,10 +8070,11 @@ function get_available_post_mime_types( $type = 'attachment' ) {
 
 	if ( ! is_array( $mime_types ) ) {
-		$mime_types = $wpdb->get_col( $wpdb->prepare( "SELECT DISTINCT post_mime_type FROM $wpdb->posts WHERE post_type = %s", $type ) );
+		$mime_types = $wpdb->get_col( $wpdb->prepare( "SELECT DISTINCT post_mime_type FROM $wpdb->posts WHERE post_type = %s AND post_mime_type != ''", $type ) );
```
It appears that the `<>` operator is included in the ANSI standard, but most SQL databases also support the `!=` operator.
Therefore, there is virtually no difference between the two.

ironprogrammer replied:
Agreed -- the reason I chose `!=` is because it's commonly used in Core's SQL queries.

anton-vlasenko reviewed (approved):
I've [tested this PR](https://core.trac.wordpress.org/ticket/59195#comment:16) and it works as described.
LGTM!

anton-vlasenko commented on src/wp-includes/post.php:
```diff
@@ -8070,10 +8070,11 @@ function get_available_post_mime_types( $type = 'attachment' ) {
 
-	return $mime_types;
+	// Remove nulls from returned $mime_types.
+	return array_filter( $mime_types );
```
As `array_filter()` preserves the keys, this may result in gaps if the array was indexed.
I'm not sure if that's an issue, it's just that I'd expect sequential keys to be returned by the `get_available_post_mime_types()` function.
Here's an example: https://3v4l.org/FIfCm

ironprogrammer replied:
It doesn't seem like that would be an issue in how the array is ultimately iterated. [In `post.php` it uses](https://github.com/WordPress/wordpress-develop/blob/204a1bbf4e5f22b07a93c1f4a0b12bdd65d6483f/src/wp-includes/post.php#L3377)
```php
foreach ( (array) $real_mime_types as $real ) {
    ...
```
where `$real_mime_types` is set from `$avail_post_mime_types`.

That said, what do you think about
```suggestion
	return array_values( array_filter( $mime_types ) );
```
to leave things tidy, especially if the keys would be expected to be sequential?

anton-vlasenko replied:
Yes, it looks good. As far as I can tell, `$mime_types` is always an array at this step. So, it should be safe to call these 2 functions.

anton-vlasenko reviewed (commented):
I'm sorry, I've just noticed this.
I'm leaving it to your discretion.

ironprogrammer committed:
Re-index array before return

hellofromtonya reviewed (approved):
`get_available_post_mime_types()` is documented to return an array of `string`s. It should not be returning `null` elements.

Thus the root cause of the deprecation notice is this function returning `null` in 1 or more elements.

Removing the empty elements makes sense as these are not valid MIME types.

Tested to confirm this patch does resolve the root cause of the deprecation notice ✅ 

Will mark for `commit`.

hellofromtonya commented on tests/phpunit/tests/post/mimeTypes.php:
```diff
@@ -0,0 +1,66 @@
+ * @covers ::get_available_post_mime_types
+ */
+
+class Tests_Get_Available_Post_Mime_Types extends WP_UnitTestCase {
```
```suggestion
class Tests_Post_GetAvailablePostMimeTypes extends WP_UnitTestCase {
```

hellofromtonya committed:
Renames test for coding standards.

hellofromtonya committed:
Rename test file for coding standards.

hellofromtonya commented:
Committed via https://core.trac.wordpress.org/changeset/58437.
